data_and_models/pipelines/ner/dvc.yaml

```diff
@@ -1,404 +1,170 @@
-      --code initialize.py
-      --paths.train ../../annotations/ner/annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.train.spacy
-      --paths.dev ../../annotations/ner/annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.dev.spacy
-      --paths.base_model en_ner_craft_md
-      &&
-      python patch.py ../../models/ner/model5/model-best/vocab/strings.json
-    deps:
-    - Dockerfile
-    - config.cfg
-    - initialize.py
-    - ../../annotations/ner/annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.train.spacy
-    - ../../annotations/ner/annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.dev.spacy
-    outs:
-    - ../../models/ner/model5/model-best/
+  convert_annotations:
+    foreach:
+      annotation5:
+        annotation_filename: annotations5_EmmanuelleLogette_2020-06-30_raw2_Disease
+      annotation6:
+        annotation_filename: annotations6_EmmanuelleLogette_2020-07-07_raw4_TaxonChebi
+      annotation9:
+        annotation_filename: annotations9_EmmanuelleLogette_2020-07-08_raw6_CelltypeProtein
+      annotation10:
+        annotation_filename: annotations10_EmmanuelleLogette_2020-08-28_raw1_raw5_10EntityTypes
+      annotation11:
+        annotation_filename: annotations11_CharlotteLorin_2020-08-28_raw1_10EntityTypes
+      annotation12:
+        annotation_filename: annotations12_EmmanuelleLogette_2020-08-28_raw7_10EntityTypes
+      annotation13:
+        annotation_filename: annotations13_CharlotteLorin_2020-09-02_raw7_10EntityTypes
+      annotation14:
+        annotation_filename: annotations14_EmmanuelleLogette_2020-09-02_raw8_CellCompartmentDrugOrgan
+      annotation15:
+        annotation_filename: annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway
+    do:
+      cmd: python preprocess.py ../../annotations/ner/${item.annotation_filename}.jsonl
+      deps:
+        - Dockerfile
+        - preprocess.py
+        - ../../annotations/ner/${item.annotation_filename}.jsonl
+      params:
+        - train.corpora.dev_size
+        - train.corpora.shuffle_seed
+      outs:
+        - ../../annotations/ner/${item.annotation_filename}.train.spacy
+        - ../../annotations/ner/${item.annotation_filename}.dev.spacy
+  training:
+    foreach:
+       model1:
+         annotation_filename: annotations5_EmmanuelleLogette_2020-06-30_raw2_Disease
+         base_model: en_ner_bc5cdr_md
+       model2:
+         annotation_filename: annotations14_EmmanuelleLogette_2020-09-02_raw8_CellCompartmentDrugOrgan
+         base_model: en_ner_bionlp13cg_md
+       model3:
+         annotation_filename: annotations6_EmmanuelleLogette_2020-07-07_raw4_TaxonChebi
+         base_model: en_ner_craft_md
+       model4:
+         annotation_filename: annotations9_EmmanuelleLogette_2020-07-08_raw6_CelltypeProtein
+         base_model: en_ner_jnlpba_md
+       model5:
+         annotation_filename: annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway
+         base_model: en_ner_craft_md
```

Comment:
I don't know if it's possible, probably it's a discussion for another PR/issue, but maybe someone has some thoughts/insight on this already.

I see that in all of the stages we use similar data about the models, their entity types, and the corresponding annotations. It feels to me that this might be better  stored in `params.yaml` (correct me if I'm wrong!) Apparently one can iterate over dictionaries from `params.yaml` in the `dvc.yaml` file, here's the example from the [DVC docs](https://dvc.org/doc/user-guide/project-structure/pipelines-files#foreach-stages):
```yaml
stages:
  mystages:
    foreach: ${myobject} # From params.yaml
    do:
      cmd: ./script.py ${key} ${item.prop1}
      outs:
        - ${item.prop2}
```
I'm just thinking out loud, but couldn't we put something like this in the `params.yaml` file and then use it in `dvc.yaml` like in the example above:
```yaml
#params.yaml
...
models:
  model1:
    annotation_filename: annotations5_EmmanuelleLogette_2020-06-30_raw2_Disease
    base_model: en_ner_bc5cdr_md
    entity_types:
      DISEASE: DISEASE
  model2:
    annotation_filename: annotations14_EmmanuelleLogette_2020-09-02_raw8_CellCompartmentDrugOrgan
    base_model: en_ner_bionlp13cg_md
    entity_types:
      CELL_COMPARTMENT: CELLULAR_COMPONENT
      DRUG: SIMPLE_CHEMICAL
      ORGAN: ORGAN
...
```

Reply:
Thanks @Stannislav for this suggestion! Maybe just some points to keep in mind for this discussion:
- Initially, `params.yaml` was created to keep track of the `hyperparameters` along the pipeline (as `n_epochs`, `dropout`, ...). There is indeed a DVC command `dvc params diff` allowing to do the comparison between several runs with different parameters. I have the feeling that the `dvc.yaml` is the file used to create the "links" between the steps/dependencies (DAG). So from your example, in my opinion, the `base model` could be a `hyperparameter`, I am not sure about the `annotation filename`. 
- I am wondering if changing one key of this dictionary `model1` could affect steps that are not changing but depends on this dictionary anyway. For example, you want to change the `base model`, the step `convert_annotation` is not depending on `base model` but its object changes. I am not sure how DVC handles this. 

In general, I also have the feeling that those structures could be re-think with the `config.cfg` introduced by `spaCy 3.0` as most of the hyper-parameters seems to be located in that file now.

Reply:
Thanks for the comments, I see now that maybe `params.yaml` might not be the best place for this, as it has a specific use of specifying hyperparameters.

Regarding changing some of the keys and re-computing the steps in the DAG. I'd assume that the structrue above is absolutely fixed and we'd never change it anyway. But yes, I also agree that for example the data conversion should not care about what model the data files are used for.

Anyway, this comment was not meant to lead to changes in this PR, just for duscussing ideas :) feel free to carry on :)